docs/abi.rst

```diff
@@ -0,0 +1,919 @@
+.. _abi:
+
+ABI Support
+===========
+
+.. warning::
+    ABI support is still taking shape and is subject to backwards incompatible changes.
+        * Based on feedback, the API and usage patterns are likely to change.
+        * For the following use cases, feel encouraged to rely on abstractions. Expect a best-effort attempt to minimize backwards incompatible changes along with a migration path.
+
+            * :any:`ABIReturnSubroutine` usage for ABI Application entry point definition.
+
+            * :any:`Router` usage for defining how to route program invocations.
+        * For general purpose :any:`Subroutine` definition usage, use at your own risk. Based on feedback, the API and usage patterns will change more freely and with less effort to provide migration paths.
+
+    For these reasons, we strongly recommend using :any:`pragma` or the :any:`Pragma` expression to pin the version of PyTeal in your source code.
+
+`ARC-4 <https://arc.algorand.foundation/ARCs/arc-0004>`_ introduces a set of standards that increase the interoperability of smart contracts in the Algorand ecosystem. This set of standards is commonly referred to as Algorand's application binary interface, or ABI.
+
+This page will introduce and explain the relevant concepts necessary to build a PyTeal application that adheres to the ARC-4 ABI standards.
+
+Types
+------
+
+The ABI supports a variety of data types whose encodings are standardized.
+
+.. note::
+    Be aware that the ABI type system in PyTeal has been designed specifically for the limited use case of describing a program's inputs and outputs. At the time of writing, we **do not recommend** using ABI types in a program's internal storage or computation logic, as the more basic :any:`TealType.uint64` and :any:`TealType.bytes` :any:`Expr` types are far more efficient for these purposes.
+
```

Comment:
This q should _not_ block release, just putting it here for consideration

What is the appropriate way to store things that come in as ABI types? for dynamic types there may be a lot of overhead in the encoded version but static types are more efficient. Should we recommend that _if_ you're storing data, it should be a static type and use `encode`?

Reply:
Yes, I don't see much of a downside for storing encoded static types.

Unless you want to store single integers in local/global state, in which case you need not convert them to bytes anyway.

For the time being I'd like to retain this warning, but perhaps over time we can clarify this point.